File: contrib/jxpmap/org/xpilot/jxpmap/MapModel.java
package org.xpilot.jxpmap;

import java.awt.Color;
import java.awt.Point;
import java.awt.Polygon;
import java.awt.Rectangle;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.io.StringReader;
import java.io.StringWriter;
import java.io.ByteArrayInputStream;
import java.io.InputStream;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.BufferedInputStream;
import java.net.URL;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.ListIterator;
import java.util.Map;
import java.math.BigDecimal;

import javax.xml.parsers.*;
import javax.xml.*;
import org.w3c.dom.*;
import org.xml.sax.*;

import javax.swing.JOptionPane;

public class MapModel extends ModelObject {

    List pixmaps;
    List objects;
    List edgeStyles;
    List polyStyles;
    MapOptions options;
    float defaultScale;
    LineStyle defEdgeStyle;
    Map defPolygonStyles;

    public Object deepClone (Map context) {

        MapModel clone = (MapModel)super.deepClone(context);
            
        ArrayList l = new ArrayList();
        for (Iterator i = pixmaps.iterator(); i.hasNext();)
            l.add(((Pixmap)i.next()).deepClone(context));
        clone.pixmaps = l;

        l = new ArrayList();
        for (Iterator i = objects.iterator(); i.hasNext();)
            l.add(((MapObject)i.next()).deepClone(context));
        clone.objects = l;
        
        l = new ArrayList();
        for (Iterator i = edgeStyles.iterator(); i.hasNext();)
            l.add(((LineStyle)i.next()).deepClone(context));
        clone.edgeStyles = l;
        
        l = new ArrayList();
        for (Iterator i = polyStyles.iterator(); i.hasNext();)
            l.add(((PolygonStyle)i.next()).deepClone(context));
        clone.polyStyles = l;
        
        clone.options = new MapOptions(options);
        return clone;
    }
    
    public MapModel () {

        defaultScale = 1f / 64;
        objects = new ArrayList();
        polyStyles = new ArrayList();
        pixmaps = new ArrayList();
        edgeStyles = new ArrayList();
        options = new MapOptions();

        LineStyle ls;
        ls = new LineStyle("internal", 0, Color.black, LineStyle.STYLE_HIDDEN);
        edgeStyles.add(ls);
        ls = new LineStyle("default", 0, Color.blue, LineStyle.STYLE_SOLID);
        edgeStyles.add(ls);
        defEdgeStyle = ls;
        
        PolygonStyle ps;
        ps = new PolygonStyle();
        ps.setId("default");
        ps.setVisible(true);
        ps.setVisibleInRadar(true);
        ps.setFillStyle(PolygonStyle.FILL_NONE);
        ps.setDefaultEdgeStyle(defEdgeStyle);
        polyStyles.add(ps);
        defPolygonStyles = new HashMap();
        defPolygonStyles.put("default", ps);
    }


    public List getEdgeStyles() {
        return edgeStyles;
    }


    public List getObjects() {
        return objects;
    }


    public MapOptions getOptions() {
        return options;
    }


    public List getPixmaps() {
        return pixmaps;
    }


    public List getPolyStyles() {
        return polyStyles;
    }


    public float getDefaultScale () {
        return defaultScale;
    }


    public void setDefaultScale (float def) {
        this.defaultScale = def;
    }


    public LineStyle getDefaultEdgeStyle () {
        return defEdgeStyle;
    }


    public void setDefaultEdgeStyle (LineStyle style) {
        this.defEdgeStyle = style;

    }


    public Map getDefaultPolygonStyles() {
        return defPolygonStyles;
    }
    
    
    public void setDefaultPolygonStyles (Map styles) {
        this.defPolygonStyles = styles;
    }


    public void addToFront (MapObject moNew) {
        addObject(moNew, true);
    }


    public void addToBack (MapObject moNew) {
        addObject(moNew, false);
    }


    public void addObject (MapObject moNew, boolean front) {

        for (ListIterator iter = objects.listIterator(); iter.hasNext();) {
            MapObject moOld = (MapObject)iter.next();
            int znew = moNew.getZOrder();
            int zold = moOld.getZOrder();
            
            if (znew < zold || (front && znew == zold)) {
                iter.previous();
                iter.add(moNew);
                return;
            }
        }
        objects.add(moNew);
    }

    public void addObject (int index, MapObject mo) {
        objects.add(index, mo);
    }
    
    public void replace (MapObject from, MapObject to) {
        int i = objects.indexOf(from);
        objects.set(i, to);
    }
    
    public int indexOf (MapObject mo) {
        return objects.indexOf(mo);
    }

    public void removeObject (MapObject mo) {
        objects.remove(mo);
    }

    public void removeObject (int index) {
        objects.remove(index);
    }
    
    public Object[] validateModel () {
        for (Iterator i = objects.iterator(); i.hasNext();) {
            MapObject o = (MapObject)i.next();
            if (o instanceof MapPolygon) {
                Polygon p = ((MapPolygon)o).getPolygon();
                if (p.npoints < 1) continue;
                int x = p.xpoints[0];
                int y = p.ypoints[0];
                if (x < 0 || x >= options.getSize().width
                || y < 0 || y >= options.getSize().height) {
                    return new Object[] {
                    o, "Polygon is located outside map bounds." };
                }                
            } else if (!(o instanceof Group)) {
                Rectangle b = o.getBounds();
                if (b.x < 0 || b.x >= options.getSize().width
                || b.y < 0 || b.y >= options.getSize().height) {
                    return new Object[] {
                    o, "Object is located outside map bounds." };
                }
            }
        }
        return null;
    }


    public void load (String name) 
    throws IOException, ParserConfigurationException,
    SAXException, ParseException, XPDFile.ParseException {
            
        edgeStyles.clear();
        LineStyle ls;
        ls = new LineStyle("internal", 0, Color.black, LineStyle.STYLE_HIDDEN);
        edgeStyles.add(ls);
        polyStyles.clear();
        pixmaps.clear();

        XPDFile xpd = XPDFile.load(new File(name));
        XPDFile.Part first = (XPDFile.Part)xpd.get(0);
        readXml(new String(first.data, "ISO8859_1"));

        for (Iterator i = pixmaps.iterator(); i.hasNext();) {
            Pixmap pixmap = (Pixmap)i.next();
            for (Iterator j = xpd.iterator(); j.hasNext();) {
                XPDFile.Part part = (XPDFile.Part)j.next();
                if (pixmap.getFileName().equals(part.name))
                    pixmap.load(new ByteArrayInputStream(part.data));
            }
            if (pixmap.getImage() == null)
                throw new IOException("missing image: " 
                    + pixmap.getFileName());
        }

        defEdgeStyle = (LineStyle)edgeStyles.get(edgeStyles.size() - 1);
        defPolygonStyles = new HashMap();
        defPolygonStyles.put("default", polyStyles.get(polyStyles.size() - 1));
    }
    
    /*
    private void readXml (String xml) 
    throws IOException, SAXException, ParserConfigurationException {
        SAXParserFactory spf = SAXParserFactory.newInstance();
        spf.setValidating(false);
        XMLReader reader = spf.newSAXParser().getXMLReader();
        reader.setContentHandler(new MapDocumentHandler());
        reader.parse(new InputSource(new StringReader(xml)));
    }
    */

    public void save (File file) throws IOException {
        
        String xml = exportXml();
        
        String xmlName = file.getName();
        int dot = xmlName.lastIndexOf('.');
        if (dot != -1) xmlName = xmlName.substring(0, dot);
        xmlName += ".xp2";
        
        XPDFile xpd = new XPDFile();
        xpd.add(new XPDFile.Part(xmlName, xml.getBytes("ISO8859_1")));
        for (Iterator i = pixmaps.iterator(); i.hasNext();) {
            Pixmap pixmap = (Pixmap)i.next();
            xpd.add(new XPDFile.Part(pixmap.getFileName(), 
                new PPMEncoder().encode(pixmap.getImage())));
        }
        
        xpd.save(file);
    }
    
    
    private String downloadImages() throws IOException {
        String urlStr = options.get("dataurl");
        if (urlStr == null) {
            JOptionPane.showMessageDialog(null,
                "The map has no dataURL option");
            return "";
        }
        File tmpDir = File.createTempFile("jxpmap", null);
        // JDK seems to create an actual file
        // I need to delete it before it can be used as a dir
        tmpDir.delete();
        if (!tmpDir.mkdir())
            throw new IOException("unable to create temporary directory: " 
                + tmpDir.getAbsolutePath());
        tmpDir.deleteOnExit();
        URL url = new URL(urlStr);
        File xpdFile = new File(tmpDir, new File(url.getPath()).getName());        
        InputStream in = url.openStream();
        try {
            FileOutputStream out = new FileOutputStream(xpdFile);
            try {
                byte buf[] = new byte[2048];
                for (int c = in.read(buf); c != -1; c = in.read(buf))
                    out.write(buf, 0, c);
            } finally {
                out.close();
            }
        } finally {
            in.close();
        }
        try {
            XPDFile xpd = XPDFile.load(xpdFile);
            xpdFile.delete();
            for (Iterator i = xpd.iterator(); i.hasNext();) {
                XPDFile.Part p = (XPDFile.Part)i.next();
                File pf = new File(tmpDir, p.name);
                FileOutputStream out = new FileOutputStream(pf);
                try {
                    out.write(p.data);
                } finally {
                    out.close();
                }
                pf.deleteOnExit();
            }
            return tmpDir.getAbsolutePath() + "/";
        } catch (XPDFile.ParseException pe) {
            throw new IOException("corrupted xpd file");
        }
    }
    
    
    public void importXml(String xml)
    throws IOException, ParseException, 
    SAXException, ParserConfigurationException {
        edgeStyles.clear();
        LineStyle ls;
        ls = new LineStyle("internal", 0, Color.black, LineStyle.STYLE_HIDDEN);
        edgeStyles.add(ls);
        polyStyles.clear();
        pixmaps.clear();

        readXml(xml);

        boolean asked = false;
        String tmpDir = "";
        for (Iterator i = pixmaps.iterator(); i.hasNext();) {
            Pixmap pixmap = (Pixmap)i.next();
            File f = new File(tmpDir + pixmap.getFileName());
            if (!asked && !f.exists()) {
                asked = true;
                if (JOptionPane.showConfirmDialog(null,
                    "Download the images used by this map?") 
                    == JOptionPane.YES_OPTION) {
                    tmpDir = downloadImages();
                    f = new File(tmpDir + pixmap.getFileName());
                }
            }
            if (f.exists()) {
                InputStream in = new BufferedInputStream(
                    new FileInputStream(f));
                try {
                    pixmap.load(in);
                } finally {
                    in.close();
                }
            }
        }

        defEdgeStyle = (LineStyle)edgeStyles.get(edgeStyles.size() - 1);
        defPolygonStyles.put("default", polyStyles.get(polyStyles.size() - 1));
    }
    
    public String exportXml() throws IOException {
        
        StringWriter w = new StringWriter();
        PrintWriter out = new PrintWriter(w);
        
        out.println("<XPilotMap version=\"1.1\">");  
        options.printXml(out);
        for (Iterator iter = pixmaps.iterator(); iter.hasNext();) {
            Pixmap p = (Pixmap)iter.next();
            p.printXml(out);
        }
        for (Iterator iter = edgeStyles.iterator(); iter.hasNext();) {
            LineStyle s = (LineStyle)iter.next();
            s.printXml(out);
        }
        for (Iterator iter = polyStyles.iterator(); iter.hasNext();) {
            PolygonStyle s = (PolygonStyle)iter.next();
            s.printXml(out);
        }
        for (int i = objects.size() - 1; i >= 0; i--) {
            ((MapObject)objects.get(i)).printXml(out);
        }
        out.println("</XPilotMap>");

        return w.toString();        
    }
    
    private String atts(Node tag, String name) {
        Node att = tag.getAttributes().getNamedItem(name);
        return att != null ? att.getNodeValue() : null;
    }
    
    private int atti(Node tag, String name) {
        return Integer.parseInt(atts(tag, name));
    }
    
    private int atti(Node tag, String name, int def) {
        String s = atts(tag, name);
        if (s == null) return def;
        return Integer.parseInt(s);
    }
    
    private Iterator tags(Node parent, String name) {
        ArrayList list = new ArrayList();
        for (Node n = parent.getFirstChild(); 
        n != null; 
        n = n.getNextSibling()) {
            if (n.getNodeType() != Node.ELEMENT_NODE) continue;
            if (!name.equalsIgnoreCase(n.getNodeName())) continue;
            list.add(n);
        }
        return list.iterator();
    }
    
    private void readXml(String xml)
    throws IOException, SAXException, 
    ParseException, ParserConfigurationException {
        Document doc = 
            DocumentBuilderFactory.newInstance().newDocumentBuilder().parse(
                new InputSource(new StringReader(xml)));
        Element root = doc.getDocumentElement();
        if (!"xpilotmap".equalsIgnoreCase(root.getNodeName()))
            throw new ParseException("expected XPilotMap");
        if (root.getFirstChild() == null) return;
        
        Map bstyles = new HashMap();
        for (Iterator i = tags(root, "bmpstyle"); i.hasNext();) {
            Node n = (Node)i.next();
            Pixmap pm = new Pixmap();
            pm.setFileName(atts(n, "filename"));
            String s = atts(n, "scalable");
            if (s != null && !"yes".equals(s))
                pm.setScalable(false);
            bstyles.put(atts(n, "id"), pm);
            pixmaps.add(pm);
        }
        
        Map estyles = new HashMap();
        for (Iterator i = tags(root, "edgestyle"); i.hasNext();) {
            Node n = (Node)i.next();
            LineStyle ls = new LineStyle();
            ls.setId(atts(n, "id"));
            ls.setWidth(atti(n, "width"));
            ls.setStyle(atti(n, "style"));
            if (ls.getWidth() == -1) {
                ls.setWidth(1);
                ls.setStyle(LineStyle.STYLE_HIDDEN);
            }
            ls.setColor(new Color(Integer.parseInt(atts(n, "color"), 16)));
            estyles.put(ls.getId(), ls);
            edgeStyles.add(ls);
        }
        
        Map pstyles = new HashMap();
        for (Iterator i = tags(root, "polystyle"); i.hasNext();) {
            Node n = (Node)i.next();
            PolygonStyle style = new PolygonStyle();                             
            style.setId(atts(n, "id"));
            String cstr = atts(n, "color");
            if (cstr != null)
                style.setColor(
                    new Color(Integer.parseInt(cstr, 16)));      
            String txid = atts(n, "texture");
            if (txid != null)
                style.setTexture((Pixmap)bstyles.get(txid));
            style.parseFlags(atti(n, "flags", 1));
            LineStyle ls = null;
            String lsid = atts(n, "defedge");
            if (lsid != null) ls = (LineStyle)estyles.get(lsid);
            if (ls == null)
                throw new ParseException(
                    "Undefined edge style: " + lsid);
            style.setDefaultEdgeStyle(ls);
            polyStyles.add(style);
            pstyles.put(style.getId(), style);            
        }
        
        Group temp = new Group();
        readObjects(temp, root, pstyles, estyles);
        objects.addAll(temp.getMembers());
    }
    
    private void 
    readObjects(Group parent, Node root, Map pstyles, Map estyles)
    throws ParseException {
        
        for (Node n = root.getFirstChild(); n != null; 
        n = n.getNextSibling()) {
            
            if (n.getNodeType() != Node.ELEMENT_NODE) continue;
            String name = n.getNodeName();
            
            if (name.equalsIgnoreCase("generaloptions")) {
                
                for (Node n2 = n.getFirstChild(); n2 != null; 
                n2 = n2.getNextSibling()) {
                    if (n2.getNodeType() != Node.ELEMENT_NODE) continue;
                    if (!"option".equalsIgnoreCase(n2.getNodeName())) {
                        System.err.println("unknown tag inside options: "
                            + n2.getNodeName());
                        continue;
                    }
                    options.set(atts(n2, "name"), atts(n2, "value"));
                }
                
            } else if (name.equalsIgnoreCase("polygon")) {
                
                PolygonStyle s = (PolygonStyle)pstyles.get(atts(n, "style"));
                if (s == null) throw new ParseException(
                    "undefined style: " + atts(n, "style")); 
                ArrayList points = new ArrayList();
                int x = atti(n, "x");
                int y = atti(n, "y");
                points.add(new Point(x, y));
                ArrayList edges = new ArrayList();
                HashMap states = new HashMap();
                states.put("default", s);
                boolean special = false;
                for (Node n2 = n.getFirstChild(); n2 != null; 
                n2 = n2.getNextSibling()) {
                    if (n2.getNodeType() != Node.ELEMENT_NODE) continue;
                    String name2 = n2.getNodeName();
                    if (name2.equalsIgnoreCase("offset")) {
                        x += atti(n2, "x");
                        y += atti(n2, "y");
                        points.add(new Point(x, y));
                        LineStyle es = null;
                        String esName = atts(n2, "style");
                        if (esName != null) 
                            es = (LineStyle)estyles.get(esName);
                        if (es != null) special = true;
                        edges.add(es);
                    } else if (name2.equalsIgnoreCase("style")) {
                        PolygonStyle ps = 
							(PolygonStyle)pstyles.get(atts(n2, "id"));
                        if (ps != null) states.put(atts(n2, "state"), ps);
                        else System.err.println(
                            "undefined state style: " + atts(n2, "id"));
                    } else {
                        System.err.println("unknown tag inside polygon: "
                        + name2);
                    }
                }
                Polygon p = new Polygon();
                for (int i = 0; i < points.size() - 1; i++) {
                    Point pnt = (Point)points.get(i);
                    p.addPoint(pnt.x, pnt.y);
                }
                MapPolygon mp = 
                    new MapPolygon(p, states, special ? edges : null);
                parent.addToFront(mp);
                
            } else if (name.equalsIgnoreCase("fuel")) {

                MapObject o = SimpleMapObject.createFuel();
                Rectangle r = o.getBounds();
                o.moveTo(atti(n, "x") - r.width / 2, 
                    atti(n, "y") - r.height / 2);
                parent.addToFront(o);

            } else if (name.equalsIgnoreCase("ball")) {

                PolygonStyle style = null;
                String styleName = atts(n, "style");
                if (styleName != null) {
                    style = (PolygonStyle)pstyles.get(styleName);
                    if (style == null)
                        System.err.println("undefined style: " 
                            + styleName);
                }
                Ball o = new Ball(
                    atti(n, "x"),
                    atti(n, "y"),
                    atti(n, "team", 0xffff),
                    style);
                Rectangle r = o.getBounds();
                o.moveTo(r.x - r.width / 2, r.y - r.height / 2);
                parent.addToFront(o);

            } else if (name.equalsIgnoreCase("base")) {

                Base o = new Base(
                    atti(n, "x"),
                    atti(n, "y"),
                    atti(n, "dir"),
                    atti(n, "team", -1),
                    atti(n, "order", 0));
                Rectangle r = o.getBounds();
                o.moveTo(r.x - r.width / 2, r.y - r.height / 2);
                parent.addToFront(o);

            } else if (name.equalsIgnoreCase("check")) {

                MapObject o = SimpleMapObject.createCheck();
                Rectangle r = o.getBounds();
                o.moveTo(atti(n, "x") - r.width / 2, 
                    atti(n, "y") - r.height / 2);
                parent.addToFront(o);
                
            } else if (name.equalsIgnoreCase("itemconcentrator")) {

                MapObject o = SimpleMapObject.createItemConcentrator();
                Rectangle r = o.getBounds();
                o.moveTo(atti(n, "x") - r.width / 2, 
                    atti(n, "y") - r.height / 2);
                parent.addToFront(o);

            } else if (name.equalsIgnoreCase("asteroidconcentrator")) {

                MapObject o = SimpleMapObject.createAsteroidConcentrator();
                Rectangle r = o.getBounds();
                o.moveTo(atti(n, "x") - r.width / 2, 
                    atti(n, "y") - r.height / 2);
                parent.addToFront(o);
                
            } else if (name.equalsIgnoreCase("grav")) {

                Grav g = new Grav(
                    atti(n, "x"),
                    atti(n, "y"),
                    atts(n, "type"),
                    atts(n, "force"));
                Rectangle r = g.getBounds();
                g.moveTo(r.x - r.width / 2, r.y - r.height / 2);
                parent.addToFront(g);

            } else if (name.equalsIgnoreCase("wormhole")) {

                Wormhole wh = new Wormhole();
                wh.setPoint(atti(n, "x"), atti(n, "y"));
                wh.setType(atts(n, "type"));
                readObjects(wh, n, pstyles, estyles);
                parent.addToFront(wh);

            } else if (name.equalsIgnoreCase("ballarea")) {
                
                BallArea ba = new BallArea();
                readObjects(ba, n, pstyles, estyles);
                parent.addToFront(ba);

            } else if (name.equalsIgnoreCase("balltarget")) {

                BallTarget bt = new BallTarget();
                bt.setTeam(atti(n, "team"));
                readObjects(bt, n, pstyles, estyles);
                parent.addToFront(bt);

            } else if (name.equalsIgnoreCase("decor")) {
                
                Decoration d = new Decoration();
                readObjects(d, n, pstyles, estyles);
                parent.addToFront(d);

            } else if (name.equalsIgnoreCase("cannon")) {
                
                Cannon c = new Cannon();
                c.setTeam(atti(n, "team", -1));
                c.setPoint(atti(n, "x"), atti(n, "y"));
                c.setDir(atti(n, "dir"));
                readObjects(c, n, pstyles, estyles);
                parent.addToFront(c);                
                
            } else if (name.equalsIgnoreCase("target")) {
                
                Target t = new Target();
                t.setTeam(atti(n, "team", -1));
                readObjects(t, n, pstyles, estyles);
                parent.addToFront(t);
                
            } else if (name.equalsIgnoreCase("frictionarea")) {
                
                FrictionArea fa = new FrictionArea();
                fa.setFriction(new BigDecimal(atts(n, "friction")));
                readObjects(fa, n, pstyles, estyles);
                parent.addToFront(fa);
                
            } else if (name.equalsIgnoreCase("group")) {
                
                Group g = new Group();
                readObjects(g, n, pstyles, estyles);
                parent.addToFront(g);
                
            } else if (name.equalsIgnoreCase("bmpstyle") 
            || name.equalsIgnoreCase("edgestyle") 
            || name.equalsIgnoreCase("polystyle")) {
                /* ignore */
            } else {
                System.err.println("unknown map element: " + name);
            }
        }
    }
    
    public static class ParseException extends Exception {
        public ParseException(String msg) {
            super(msg);
        }
    }    
}

File: contrib/jxpmap/python/MapModel.py
from __future__ import nested_scopes
from LineStyle import LineStyle
from PolygonStyle import PolygonStyle
from MapOptions import MapOptions
from MapBall import MapBall
from MapPolygon import MapPolygon
from MapBase import MapBase
from MapFuel import MapFuel
from MapCheckPoint import MapCheckPoint
from Pixmap import Pixmap
from javastuff import *
import xml.sax

class Poly:
    def __init__(self):
        self. hasSpecialEdges = 0
        self.points = []
        self.type = MapPolygon.TYPE_NORMAL
        self.team = 1

class PolyPoint:
    def __init__(self, x, y, style):
        self.x = int(x)
        self.y = int(y)
        self.style = style

class PolyStyle:
    def __init__(self, id, color, textureId, defEdgeId, flags):
        self.id = id
        self.color = color
        self.textureId = textureId
        self.defEdgeId = defEdgeId
        self.flags = flags

class MapModel:
    def __init__(self):
        self.defaultScale = 1./64
        self.objects = []
        self.pixmaps = []
        self.polyStyles = []
        self.edgeStyles = []
        self.options = MapOptions()

        ls = LineStyle("internal", 0, 0xFFFF, LineStyle.STYLE_INTERNAL)
        self.edgeStyles.append(ls)
        ls = LineStyle("default", 0, 0xFF, LineStyle.STYLE_SOLID)
        self.edgeStyles.append(ls)
        self.defEdgeStyleIndex = 1
        ps = PolygonStyle()
        ps.id = "default"
        ps.visible = ps.visibleInRadar= 1
        ps.filled = ps.textured = 0
        ps.defaultEdgeStyle = ls
        self.polyStyles.append(ps)
        self.defPolygonStyleIndex = 1

    def getDefaultScale(self):
        return self.defaultScale

    def setDefaultScale(self, ds):
        self.defaultScale = ds

    def getDefaultPolygonStyle(self):
        if self.defPolygonStyleIndex >= len(self.polyStyles):
            self.defPolygonStyleIndex = len(self.polyStyles) - 1
        return self.polyStyles[self.defPolygonStyleIndex]

    def setDefaultPolygonStyle(self, index):
        self.defPolygonStyleIndex = index

    def load(self, file, bmstore):
        self.edgeStyles = []
        ls = LineStyle("internal", 0, 0xFFFF, LineStyle.STYLE_INTERNAL)
        self.edgeStyles.append(ls)
        self.polyStyles = []

        xml.sax.parse(file, MapDocumentHandler(self, bmstore))
        self.defEdgeStyleIndex = len(self.edgeStyles) - 1
        self.defPolygonStyleIndex = len(self.polyStyles) - 1

    def save(self, file):
        file.write('<XPilotMap>\n')
        self.options.printXml(file)
        for pm in self.pixmaps:
            pm.printXml(file)
        for es in self.edgeStyles:
            es.printXml(file)
        for ps in self.polyStyles:
            ps.printXml(file)
        o = self.objects[:]
        o.reverse()
        for obj in o:
            obj.printXml(file)
        file.write('</XPilotMap>\n')
        file.close()

    def addToFront(self, moNew):
        self.addObject(moNew, 1)

    def addToBack(self, moNew):
        self.addObject(moNew, 0)

    def addObject(self, moNew, front):
        znew = moNew.getZOrder()
        for i in range(len(self.objects)):
            zold = self.objects[i].getZOrder()
            if znew < zold or front and znew == zold:
                self.objects.insert(i, moNew)
                return
        self.objects.append(moNew)

    def removeObject(self, mo):
        self.objects.remove(mo)

class MapParseError(Exception):
    pass

class MapDocumentHandler(xml.sax.ContentHandler):
    def __init__(self, model, bmstore):
        self.model = model
        self.bmstore = bmstore
        self.polys = []
        self.estyles = {}
        self.estyles["internal"] = model.edgeStyles[0]
        self.pstyles = {}
        self.bstyles = {}
        self.opMap = {}
        self.ballArea = 0
        self.ballTargetTeam = -1
        self.decor = 0

    def startElement(self, name, atts):
        bm = self.bmstore
        name = name.lower()
        def getxy():
            return int(atts["x"]), int(atts["y"])
        def getint(name, base = 10):
            return int(atts[name], base)
        def addobj(o):
            self.model.addToFront(o)
        if name == "polygon":
            poly = Poly()
            poly.style = atts["style"]
            poly.points.append(PolyPoint(atts["x"], atts["y"], None))
            if self.ballArea:
                poly.type = MapPolygon.TYPE_BALLAREA
            elif self.ballTargetTeam != -1:
                poly.type = MapPolygon.TYPE_BALLTARGET
                poly.team = self.ballTargetTeam
            elif self.decor:
                poly.type = MapPolygon.TYPE_DECORATION
            self.poly = poly
        elif name == "offset":
            es = atts.get("style")
            if es != None:
                self.poly.hasSpecialEdges = 1
            self.poly.points.append(PolyPoint(atts["x"], atts["y"], es))
        elif name == 'edgestyle':
            id = atts["id"]
            color = getint("color", 16)
            width = getint("width")
            style = getint("style")
            if width == -1:
                width = 0
                style = LineStyle.STYLE_HIDDEN
            ls = LineStyle(id, width, color, style)
            self.estyles[id] = ls
            self.model.edgeStyles.append(ls)
        elif name == 'polystyle':
            id = atts["id"]
            cstr = atts.get("color")
            if cstr is not None:
                col = int(cstr, 16)
            else:
                col = None
            flagstr = atts.get('flags', '1')
            flags = int(flagstr)
            self.pstyles[id] = PolyStyle(id, col, atts.get("texture"),
                                         atts["defedge"], flags)
        elif name == 'bmpstyle':
            id = atts["id"]
            fileName = atts["filename"]
            flags = getint("flags")
            pm = Pixmap(self.bmstore, fileName)
            pm.fileName = fileName
            pm.scalable = flags != 0
            self.bstyles[id] = pm
            self.model.pixmaps.append(pm)
        elif name == 'option':
            self.opMap[atts["name"]] = atts["value"]
        elif name == 'fuel':
            x, y = getxy()
            addobj(MapFuel(x, y, bm))
        elif name == 'ball':
            x, y = getxy()
            team = getint("team")
            addobj(MapBall(x, y, team, bm))
        elif name == 'base':
            x, y = getxy()
            dir = getint("dir")
            team = getint("team")
            addobj(MapBase(x, y, dir, team, bm))
        elif name == 'check':
            x,y = getxy()
            addobj(MapCheckPoint(x, y, bm))
        elif name == 'ballarea':
            self.ballArea = 1
        elif name == 'balltarget':
            self.ballTargetTeam = getint("team")
        elif name == 'decor':
            decor = 1
        elif name == 'xpilotmap':
            pass
        elif name == 'generaloptions':
            pass
        else:
            print "Tag: ", name
            raise MapParseError("Unknown tag")

    def endElement(self, name):
        name = name.lower()
        if name == 'polygon':
            self.polys.append(self.poly)
        elif name == 'ballarea':
            self.ballArea = 0
        elif name == 'balltarget':
            self.ballTargetTeam = -1
        elif name == 'decor':
            self.decor = 0

    def endDocument(self):
        self.model.options = MapOptions(self.opMap)
        for ps in self.pstyles.values():
            style = PolygonStyle()
            style.id = ps.id
            style.setFromFlags(ps.flags)
            if style.textured:
                if ps.textureId is None:
                    raise MapParseError("Textured polygon, no texture name")
                style.texture = self.bstyles[ps.textureId]
            if ps.color is None:
                style.color = 0xFFFFFF
            else:
                style.color = ps.color
            ls = self.estyles[ps.defEdgeId]
            style.defaultEdgeStyle = ls
            self.model.polyStyles.append(style)
            ps.ref = style
        for p in self.polys:
            ps = self.pstyles[p.style]
            style = ps.ref
            defls = style.defaultEdgeStyle
            if p.hasSpecialEdges:
                edges = []
            else:
                edges = None
            awtp = Polygon()
            pnt = p.points[0]
            x = pnt.x
            y = pnt.y
            awtp.addPoint(x, y)
            for pnt in p.points[1:-1]:
                x += pnt.x
                y += pnt.y
                awtp.addPoint(x, y)
            cls = None
            if edges != None:
                for pnt in p.points[1:]:
                    if pnt.style != None:
                        cls = self.estyles[pnt.style]
                        if cls == defls:
                            cls = None
                    edges.append(cls)
            mp = MapPolygon(awtp, style, edges)
            mp.type = p.type
            mp.team = p.team
            self.model.addToFront(mp)
